crates/matrix-sdk/src/client/builder.rs

```diff
@@ -371,63 +402,88 @@ impl ClientBuilder {
+                    if http_client
+                        .send(
+                            get_supported_versions::Request::new(),
+                            Some(RequestConfig::short_retry()),
+                            homeserver_url.to_string(),
+                            None,
+                            &[MatrixVersion::V1_0],
+                            Default::default(),
+                        )
+                        .await
+                        .is_ok()
+                    {
+                        result = Ok((homeserver_url.to_string(), None));
+                    } else {
+                        // If it isn't, we'll throw the discovery error instead.
+                    }
                 }
 
-                url
+                result?
             }
-            HomeserverConfig::ServerName { server: server_name, protocol } => {
-                debug!("Trying to discover the homeserver");
-
-                let homeserver = match protocol {
-                    UrlScheme::Http => format!("http://{server_name}"),
-                    UrlScheme::Https => format!("https://{server_name}"),
-                };
-
-                let well_known = http_client
-                    .send(
-                        discover_homeserver::Request::new(),
-                        Some(RequestConfig::short_retry()),
-                        homeserver,
-                        None,
-                        &[MatrixVersion::V1_0],
-                        Default::default(),
-                    )
-                    .await
-                    .map_err(|e| match e {
-                        HttpError::Api(err) => ClientBuildError::AutoDiscovery(err),
-                        err => ClientBuildError::Http(err),
-                    })?;
-
-                #[cfg(feature = "experimental-oidc")]
-                {
-                    authentication_server_info = well_known.authentication;
-                    allow_insecure_oidc = matches!(protocol, UrlScheme::Http);
-                }
-
-                #[cfg(feature = "experimental-sliding-sync")]
-                if let Some(proxy) = well_known.sliding_sync_proxy.map(|p| p.url) {
-                    sliding_sync_proxy = Url::parse(&proxy).ok();
-                }
+        };
 
-                debug!(
-                    homeserver_url = well_known.homeserver.base_url,
-                    "Discovered the homeserver"
-                );
+        #[allow(unused_variables)]
+        if let Some(well_known) = well_known {
+            #[cfg(feature = "experimental-oidc")]
+            {
+                authentication_server_info = well_known.authentication;
+            }
 
-                well_known.homeserver.base_url
+            #[cfg(feature = "experimental-sliding-sync")]
+            if let Some(proxy) = well_known.sliding_sync_proxy.map(|p| p.url) {
+                sliding_sync_proxy = Url::parse(&proxy).ok();
             }
```

Comment:
A small adjustment that'll keep the manually set value if the well-known one isn't a valid URL:

```suggestion
            if let Some(proxy) = well_known.sliding_sync_proxy.and_then(|p| Url::parse(p.url).ok()) {
                sliding_sync_proxy = proxy;
            }
```

Reply:
Oh interesting, actually this logic feels wrong to me. If I build a client with a custom proxy, I would expect that to always be used, no-matter if one was discovered or not. Will move things around and add an extra test here.